### PR TITLE
Use docker image reference from ImageStream 

### PR DESCRIPTION
**Pull request by dmage**
[BZ 1433232](https://bugzilla.redhat.com/show_bug.cgi?id=1433232)

### CODE REVIEW

**Comment by dmage** [comment]
[test]

**Comment by miminar** [comment]
The review is in progress. There's too many unrelated changes though so it will take me some time. Please focus on closely related things in a single PR next time.

**Comment by dmage** [comment]
I can open PR for each commit if you wish. But each commit depends on the previous one.

File: pkg/dockerregistry/server/repositorymiddleware.go [hunk 1]
```diff
@@ -395,6 +395,10 @@ func (r *repository) getImageOfImageStream(dgst digest.Digest) (*imageapi.Image,
 		return nil, nil, wrapKStatusErrorOnGetImage(r.name, dgst, err)
 	}
 
+	// We are going to pretend that we've got the image from the same source
+	// as the client.
+	image.DockerImageReference = tagEvent.DockerImageReference
```
Comment (miminar) [inline pkg/dockerregistry/server/repositorymiddleware.go]: I don't think we can afford this. Occasionally, registry updates the image (e.g. when it removes its manifest). We don't want to change the original `dockerImageReference` as well.

I don't advocate against using `tagEvent.DockerImageReference`. It's correct for most of the cases, but I don't think we can set it on the image object itself.

File: pkg/dockerregistry/testutil/util.go [hunk 2]
```diff
@@ -58,68 +58,50 @@ func UploadTestBlobFromReader(
...
+	}
+
+	return desc, payload, nil
 }
 
-// createRandomTarFile creates a random tarfile, returning it as an io.ReadSeeker along with its digest. An
-// error is returned if there is a problem generating valid content. Inspired by
-// github.com/vendor/docker/distribution/testutil/tarfile.go.
-func CreateRandomTarFile() (rs io.ReadSeeker, dgst digest.Digest, err error) {
+// CreateRandomTarFile creates a random tarfile and returns it's content.
```
Comment (dmage) [inline pkg/dockerregistry/testutil/util.go]: fixed

File: pkg/dockerregistry/testutil/fakeopenshift.go [hunk 3]
```diff
@@ -37,75 +38,80 @@ func NewFakeOpenShift() *FakeOpenShift {
...
-		return nil, errors.NewAlreadyExists(unversioned.GroupResource{
-			Group:    "",
-			Resource: "images",
-		}, image.Name)
+		return nil, errors.NewAlreadyExists(api.Resource("images"), image.Name)
 	}
 
-	os.logger.Debug("(*FakeOpenShift).CreateImage: ", image.Name)
-	os.imagesStorage[image.Name] = image
+	fos.logger.Debug("(*FakeOpenShift).CreateImage: ", image.Name)
```
Comment (dmage) [inline pkg/dockerregistry/testutil/fakeopenshift.go]: Method logging is done within `fos.log`. This is logging of modification of `imagesStorage`/`imageStreamsStorage`. I changed a bit log message to make it clearer.

File: pkg/dockerregistry/server/repositorymiddleware.go (hunk 1 above)
Comment (dmage) [inline pkg/dockerregistry/server/repositorymiddleware.go]: I found only one place where it was possible. Now it's fixed and covered by the test.

File: pkg/dockerregistry/server/repositorymiddleware.go [hunk 4]
```diff
@@ -372,29 +372,45 @@ func (r *repository) getImage(dgst digest.Digest) (*imageapi.Image, error) {
...
-		return nil, nil, wrapKStatusErrorOnGetImage(r.name, dgst, err)
+		return nil, nil, nil, wrapKStatusErrorOnGetImage(r.name, dgst, err)
+	}
+
+	return image, tagEvent, stream, nil
+}
+
+// getImageOfImageStream retrieves the Image with digest `dgst` for
+// the ImageStream associated with r.
+func (r *repository) getImageOfImageStream(dgst digest.Digest) (*imageapi.Image, *imageapi.ImageStream, error) {
```
Comment (miminar) [inline pkg/dockerregistry/server/repositorymiddleware.go]: Add a fat warning that the returned image doesn't contain original `dockerImageReference` and thus it's not suitable for sending back to master API.

Also describe what is the difference between this and the method above.

**Comment by dmage** [comment]
Flake #13737, re[test].

**Comment by legionus** [comment]
[merge]

**Comment by mfojtik** [comment]
[merge]

seems to be an installer issue.

**Comment by dmage** [comment]
Flake #13814

**Comment by legionus** [comment]
[merge]

**Comment by dmage** [comment]
Flake #13846

**Comment by dmage** [comment]
Actually this is a bug, we have to wait until it is fixed. :(

**Comment by legionus** [comment]
[merge]

**Comment by legionus** [comment]
[merge]

**Comment by legionus** [comment]
[merge]

**Comment by dmage** [comment]
Flake #10773

**Comment by legionus** [comment]
[merge]

**Pull request merged**